### DISCUSSION TITLE
the trait `for<'a> tower_service::Service<IncomingStream<'a>>` is not implemented for `Router<S>`

### DISCUSSION THREAD

**Discussion by buildwithzephyr** [tokio-rs/axum#2564] [Q&A] [answered]
### Summary

I can't get `axum::serve` to work on a `Router` that includes state. It tells me that a trait bound is not satisfied.

Here is a minimal example:
```rust
use std::sync::Arc;
use axum::{extract::State, routing::get, Router};

#[derive(Clone)]
struct Struct { }

async fn handler(State(_s): State<Arc<Struct>>) {}

#[tokio::main]
async fn main() {
    let state = Arc::new(Struct { });
    let app: Router<Arc<Struct>> = Router::new().route("/", get(handler)).with_state(state.clone());

    let listener = tokio::net::TcpListener::bind("0.0.0.0:3000").await.unwrap();
    let serve_future = axum::serve(listener, app);
}
```

The error that I get is that `app` cannot be used as a `Service`.
```
error[E0277]: the trait bound `for<'a> Router<Arc<Struct>>: tower_service::Service<IncomingStream<'a>>` is not satisfied
  --> src/main.rs:16:46
   |
16 |     let serve_future = axum::serve(listener, app);
   |                        -----------           ^^^ the trait `for<'a> tower_service::Service<IncomingStream<'a>>` is not implemented for `Router<Arc<Struct>>`
   |                        |
   |                        required by a bound introduced by this call
   |
   = help: the following other types implement trait `tower_service::Service<Request>`:
             <Router as tower_service::Service<axum::http::Request<B>>>
             <Router as tower_service::Service<IncomingStream<'_>>>
note: required by a bound in `serve`
  --> /Users/.../.cargo/registry/src/index.crates.io-6f17d22bba15001f/axum-0.7.4/src/serve.rs:98:8
   |
96 | ...fn serve<M, S>(tcp_listener: TcpListener, make_service: M) -> Serve<M, S>
   |       ----- required by a bound in this function
97 | ...e
98 | ...M: for<'a> Service<IncomingStream<'a>, Error = Infallible, Response = S>,
   |       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `serve`
```

If I try `app.as_service()`, or `app.into_service()`, I get similar trait bound errors.

### axum version

0.7.4

**Comment by jplatte** [comment, marked as answer]
You annotated the `app` binding with a type of `Router<Arc<Struct>>`. The (first) type parameter is for the state that needs to be supplied for the routes, `.with_state` *changes that type parameter* to any other state type that any routes added later might need. Before the router can be passed to `serve`, it must be transformed to a `Router<()>`.

You should be able to fix the problem by removing the type annotation, as in

```rust
let app = Router::new().route("/", get(handler)).with_state(state.clone());
```

**Reply by patrickelectric** [reply]
I'm having the same problem. I have a configurable router, so the type annotation is required in my case. Is there other way around it ?

**Reply by jplatte** [reply]
If you need a type annotation, make it just `Router` so the state defaults to `()`.

**Reply by ababo** [reply]
@jplatte Still need some clarification here. I need to have a function that constructs a router to be later passed to `serve()`, so the returned router has to be annotated. How can I do that if when I add state (I have one) I get `Router<MyState>` instead of `Router<()>`?

**Reply by ababo** [reply]
Does it mean `Router<MyState>` can be implicitly converted to `Router<()>` so my function should return `Router` instead?

**Reply by jplatte** [reply]
Have you read the documentation of [`Router::with_state`](https://docs.rs/axum/latest/axum/struct.Router.html#method.with_state)? It should answer your questions.

**Reply by ababo** [reply]
Thank you, now it's clear.

**Reply by daniel1302** [reply]
We should probably add this to the docs. Into the State section. I was looking for that error in docs but It was hard to find solution there.
